### PR TITLE
Delete watcher legacy templates

### PR DESCRIPTION
**Pull request by andreidan**
This adds a bit of infrastructure to be able to remove multiple legacy templates 
with one cluster state update and removes the legacy watcher index templates.

Relates to #80853

### CODE REVIEW

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/WatcherLifeCycleService.java [hunk 1]
```diff
@@ -41,14 +48,21 @@
...
+    public static final Set<String> LEGACY_WATCHER_INDEX_TEMPLATES = Set.of(".watches", ".triggered_watches");
+
     private final AtomicReference<WatcherState> state = new AtomicReference<>(WatcherState.STARTED);
     private final AtomicReference<List<ShardRouting>> previousShardRoutings = new AtomicReference<>(Collections.emptyList());
     private volatile boolean shutDown = false; // indicates that the node has been shutdown and we should never start watcher after this.
+    private final ClusterService clusterService;
     private volatile WatcherService watcherService;
     private final EnumSet<WatcherState> stopStates = EnumSet.of(WatcherState.STOPPED, WatcherState.STOPPING);
+    private final AtomicBoolean legacyTemplatesDeleteInProgress = new AtomicBoolean(false);
+    private final AtomicBoolean checkForLegacyTemplates = new AtomicBoolean(true);
```
Comment (martijnvg) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/WatcherLifeCycleService.java:61]: Should checking whether a removal of templates is in progress be part of the functionality that `MetadataIndexTemplateService.removeTemplates(...)` method offers?

File: x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/WatcherLifeCycleService.java [hunk 2]
```diff
@@ -96,6 +110,39 @@ public void clusterChanged(ClusterChangedEvent event) {
...
+                    legacyTemplatesDeleteInProgress.set(false);
+                    logger.debug(
+                        new ParameterizedMessage(
+                            "unable to delete legacy Watcher index templates [{}]",
+                            String.join(",", LEGACY_WATCHER_INDEX_TEMPLATES)
+                        ),
+                        e
+                    );
+                }));
+            }
```
Comment (martijnvg) [inline x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/WatcherLifeCycleService.java:143]: Should `checkForLegacyTemplates` be set to false when `existingTemplatesToDelete` list is empty?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java [hunk 3]
```diff
@@ -177,6 +178,76 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         );
     }
 
+    /**
+     * Removes the provided legacy index templates if they exist. This only matches the templates using exact name, wildcards are not
+     * supported.
+     *
+     * The provided templates must exist in the cluster, otherwise an {@link IndexTemplateMissingException} is reported.
+     */
+    public static void removeTemplates(
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:187]: maybe rename to `removeLegacyTemplates`? To emphasise this batch removal method is for legacy templates only.

**Review by martijnvg — COMMENTED** [review]
Looks good, I left a few comments.

**Comment by andreidan** [comment]
Superseded by #80937

**Pull request closed without merging**